File: frontend/src/types/insight.ts
/**
 * TypeScript types for Insights and Actions
 * Aligned with database schema (insights, recommended_actions tables)
 */

export type InsightType =
    | "performance"
    | "optimization"
    | "alert"
    | "opportunity"
    | "anomaly";

export type InsightSeverity = "low" | "medium" | "high" | "critical";

export type ActionStatus =
    | "pending"
    | "approved"
    | "executed"
    | "dismissed"
    | "failed";

export interface RecommendedAction {
    id: string;
    insight_id?: string;
    org_id: string;
    action_type: string;
    platform: string;
    account_id?: string;
    campaign_id?: string;
    entity_type?: string;
    entity_id?: string;
    title: string;
    description: string;
    rationale?: string;
    expected_impact?: string;
    is_executable: boolean;
    api_payload?: Record<string, unknown>;
    priority: number;
    recommended_by?: string;
    status: ActionStatus;
    executed_at?: string;
    executed_by?: string;
    execution_result?: Record<string, unknown>;
    created_at: string;
}

export interface Insight {
    id: string;
    org_id: string;
    insight_type: InsightType;
    severity: InsightSeverity;
    category?: string;
    platform?: string;
    account_id?: string;
    campaign_id?: string;
    entity_type?: string;
    entity_id?: string;
    title: string;
    summary: string;
    detailed_analysis?: string;
    metric_data?: Record<string, unknown>;
    comparison_period?: Record<string, unknown>;
    is_read: boolean;
    is_dismissed: boolean;
    is_actioned: boolean;
    read_at?: string;
    actioned_at?: string;
    ai_model?: string;
    ai_confidence?: number;
    created_at: string;
    expires_at?: string;
    recommended_actions: RecommendedAction[];
}

export interface InsightList {
    insights: Insight[];
    total: number;
    unread_count: number;
}

export interface DigestSection {
    title: string;
    content: string;
    metrics?: Record<string, unknown>;
}

export interface DailyDigest {
    id: string;
    org_id: string;
    digest_date: string;
    title?: string;
    summary: string;
    total_spend: number;
    total_impressions: number;
    total_clicks: number;
    total_conversions: number;
    currency: string;
    spend_change_percent?: number;
    impressions_change_percent?: number;
    conversions_change_percent?: number;
    sections: DigestSection[];
    insight_ids: string[];
    sent_via?: string;
    sent_at?: string;
    created_at: string;
}

export interface DigestList {
    digests: DailyDigest[];
    total: number;
}

File: backend/app/models/insight.py
"""
Ad Platform MVP - Insight Models

Pydantic models for AI-generated insights and actions.
Aligned with database schema (insights, recommended_actions tables).
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.account import Platform


class InsightType(str, Enum):
    """Types of insights (maps to insight_type column)."""
    PERFORMANCE = "performance"
    OPTIMIZATION = "optimization"
    ALERT = "alert"
    OPPORTUNITY = "opportunity"
    ANOMALY = "anomaly"


class InsightSeverity(str, Enum):
    """Insight severity levels (maps to severity column)."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ActionStatus(str, Enum):
    """Status of recommended actions."""
    PENDING = "pending"
    APPROVED = "approved"
    EXECUTED = "executed"
    DISMISSED = "dismissed"
    FAILED = "failed"


# ===========================================
# RECOMMENDED ACTION MODELS
# ===========================================

class ActionResponse(BaseModel):
    """Response model for recommended action. Aligned with recommended_actions table."""
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: str
    insight_id: Optional[str] = None
    org_id: str
    action_type: str
    platform: str
    account_id: Optional[str] = None
    campaign_id: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    title: str
    description: str
    rationale: Optional[str] = None
    expected_impact: Optional[str] = None
    is_executable: bool = False
    api_payload: Optional[dict] = None
    priority: int = 50
    recommended_by: Optional[datetime] = None
    status: str = "pending"
    executed_at: Optional[datetime] = None
    executed_by: Optional[str] = None
    execution_result: Optional[dict] = None
    created_at: datetime


class ActionList(BaseModel):
    """List of actions."""
    actions: list[ActionResponse]
    total: int
    pending_count: int


class ActionExecuteRequest(BaseModel):
    """Request to execute action."""
    confirm: bool = True


class ActionExecuteResponse(BaseModel):
    """Response after action execution."""
    success: bool
    action_id: str
    message: str
    execution_result: Optional[dict] = None


class ActionDismissRequest(BaseModel):
    """Request to dismiss action."""
    reason: Optional[str] = None


# ===========================================
# INSIGHT MODELS
# ===========================================

class InsightResponse(BaseModel):
    """Response model for insight. Aligned with insights table."""
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: str
    org_id: str
    insight_type: str = Field(..., alias="insight_type")
    severity: str = Field(..., alias="severity")
    category: Optional[str] = None
    platform: Optional[str] = None
    account_id: Optional[str] = None
    campaign_id: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    title: str
    summary: str
    detailed_analysis: Optional[str] = None
    metric_data: Optional[dict] = Field(None, alias="metric_data")
    comparison_period: Optional[dict] = None

    # Status
    is_read: bool = False
    is_dismissed: bool = False
    is_actioned: bool = False
    read_at: Optional[datetime] = None
    actioned_at: Optional[datetime] = None

    # AI metadata
    ai_model: Optional[str] = None
    ai_confidence: Optional[Decimal] = None

    # Timestamps
    created_at: datetime
    expires_at: Optional[datetime] = None

    # Related actions
    recommended_actions: list[ActionResponse] = []


class InsightList(BaseModel):
    """List of insights."""
    insights: list[InsightResponse]
    total: int
    unread_count: int


class InsightReadRequest(BaseModel):
    """Request to mark insight as read."""
    pass


class InsightDismissRequest(BaseModel):
    """Request to dismiss insight."""
    reason: Optional[str] = None


# ===========================================
# DAILY DIGEST MODELS
# ===========================================

class DigestSection(BaseModel):
    """A section in daily digest."""
    title: str
    content: str
    metrics: Optional[dict] = None


class DailyDigestResponse(BaseModel):
    """Daily digest response."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    org_id: str
    digest_date: str
    title: Optional[str] = None
    summary: str

    total_spend: Decimal = Decimal("0.00")
    total_impressions: int = 0
    total_clicks: int = 0
    total_conversions: Decimal = Decimal("0")
    currency: str = "TRY"

    spend_change_percent: Optional[Decimal] = None
    impressions_change_percent: Optional[Decimal] = None
    conversions_change_percent: Optional[Decimal] = None

    sections: list[DigestSection] = []
    insight_ids: list[str] = []

    sent_via: Optional[str] = None
    sent_at: Optional[datetime] = None
    created_at: datetime


class DigestList(BaseModel):
    """List of digests."""
    digests: list[DailyDigestResponse]
    total: int
